File: tracora/tracora/report/assets_by_department/assets_by_department.py
# Copyright (c) 2026, Kishore and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
	filters = filters or {}
	columns = get_columns()
	data = get_data(filters)
	return columns, data

def get_columns():
	return [
		{"label": _("Department"), "fieldname": "department", "fieldtype": "Data", "width": 160},
		{"label": _("Asset Tag"), "fieldname": "asset_tag", "fieldtype": "Data", "width": 130},
		{"label": _("Asset Name"), "fieldname": "asset_name", "fieldtype": "Data", "width": 180},
		{"label": _("Brand"), "fieldname": "brand", "fieldtype": "Link", "options": "Tracora Brand", "width": 120},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 120},
		{"label": _("Assigned To"), "fieldname": "assigned_to", "fieldtype": "Link", "options": "Tracora Employee", "width": 140},
	]

def get_data(filters):
	conditions = []
	values = {}

	if filters.get("department"):
		conditions.append("e.department = %(department)s")
		values["department"] = filters["department"]
	if filters.get("holding_company"):
		conditions.append("a.holding_company = %(holding_company)s")
		values["holding_company"] = filters["holding_company"]

	where_clause = (" WHERE " + " AND ".join(conditions)) if conditions else ""

	query = f"""
		SELECT
			COALESCE(e.department, 'Unassigned Stock') AS department,
			a.asset_tag,
			a.asset_name,
			a.brand,
			a.status,
			a.assigned_to
		FROM `tabTracora Asset` a
		LEFT JOIN `tabTracora Employee` e ON a.assigned_to = e.name
		{where_clause}
		ORDER BY CASE WHEN e.department IS NULL THEN 1 ELSE 0 END, e.department ASC, a.name ASC
	"""
	return frappe.db.sql(query, values, as_dict=True)

File: tracora/tracora/report/assets_by_department/assets_by_department.js
// Copyright (c) 2026, Kishore and contributors
// For license information, please see license.txt

frappe.query_reports["Assets by Department"] = {
	filters: [
		{
			fieldname: "department",
			label: __("Department"),
			fieldtype: "Link",
			options: "Tracora Department"
		},
		{
			fieldname: "holding_company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Tracora Company"
		}
	],
	formatter: function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "status" && data) {
			const colorMap = {
				"In Store": "blue",
				"Assigned": "green",
				"Under Maintenance": "orange",
				"Recovered": "cyan",
				"Retired": "gray",
				"Disposed": "darkgray",
				"Lost": "red",
				"In Transit": "purple"
			};
			const color = colorMap[data.status] || "gray";
			return `<span class="indicator-pill ${color}">${data.status}</span>`;
		}
		if (column.fieldname === "department" && data && data.department === "Unassigned Stock") {
			return `<span style="font-weight: bold; color: #718096;">Unassigned Stock</span>`;
		}
		return value;
	}
};
